keps/sig-storage/3314-csi-changed-block-tracking/README.md

```diff
@@ -0,0 +1,1692 @@
+- `audience`<br>
+  Specifies the audience string value expected in an audience-scoped authentication token
+  presented to the [sidecar](#the-external-snapshot-metadata-sidecar) by a Kubernetes client.
+  The value should be unique to the service if possible; for example, it could be the DNS name of the service.
+
+The full Custom Resource Definition is shown below:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.11.1
+    api-approved.kubernetes.io: unapproved
+  creationTimestamp: null
+  name: snapshotmetadataservices.cbt.storage.k8s.io
+spec:
+  group: cbt.storage.k8s.io
+  names:
+    kind: SnapshotMetadataService
+    listKind: SnapshotMetadataServiceList
+    plural: snapshotmetadataservices
+    singular: snapshotmetadataservice
+  scope: Cluster
+  versions:
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        description: 'The presence of a SnapshotMetadataService CR advertises the existence of a CSI
+          driver's Kubernetes SnapshotMetadata gRPC service.
+          An audience scoped Kubernetes authentication bearer token must be passed in the
+          "security_token" field of each gRPC call made by a Kubernetes backup client.'
+        properties:
+          apiVersion:
+            description: 'APIVersion defines the versioned schema of this representation
+              of an object. Servers should convert recognized schemas to the latest
+              internal value, and may reject unrecognized values.
+              More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+            type: string
+          kind:
+            description: 'Kind is a string value representing the REST resource this
+              object represents. Servers may infer this from the endpoint the client
+              submits requests to. Cannot be updated. In CamelCase.
+              More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
+            type: string
+          metadata:
+            type: object
+          spec:
+            description: This contains data needed to connect to a Kubernetes SnapshotMetadata gRPC service.
+            properties:
+              address:
+                type: string
+                description: The TCP endpoint address of the gRPC service.
+              audience:
+                type: string
+                description: The audience string value expected in a client's authentication token passed in the
+                  "security_token" field of each gRPC call.
+              caCert:
+                description: Certificate authority bundle needed by the client to validate the service.
+                format: byte
+                type: string
+            type: object
+        type: object
+    served: true
+    storage: true
+```
```

Comment:
What components are expected to be responsible for creating/updating instances of the `SnapshotMetadataService CR`?  Perhaps an operator that installs the CSI driver?

Reply:
You can include this CR in the CSI driver installation path (operator, helm etc.). For the `address` field, consider using the service DNS, instead of pod IP.

Reply:
Was mostly thinking about how certificate rotation would be handled.  The CR will have to be updated.  Or maybe it would be better to have the CR refer to a configmap containing the CA bundle which can be updated independently?

Reply:
Exactly, this KEP didn't want to dictate how user will manage their PKI resources; following the `caBundle` approach in the `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` were exactly what we had in mind.